### PR TITLE
wrapper tf.function when creating an eval dataset

### PR DESCRIPTION
**Pull request by QiJune**
Has tested with deepFM model, in CPU.

The [issue](https://github.com/tensorflow/tensorflow/issues/35084) has been opened in tensorflow.

Memory is descreased from 524787712 to 346025984

- test scripts

```bash
python -m elasticdl.python.elasticdl.client train \
  --image_base=elasticdl:ci \
  --docker_base_url=${DOCKER_BASE_URL} \
  --docker_tlscert=${DOCKER_TLSCERT} \
  --docker_tlskey=${DOCKER_TLSKEY} \
  --model_zoo=model_zoo \
  --model_def=deepfm_functional_api.deepfm_functional_api.custom_model \
  --training_data=/data/frappe/train \
  --validation_data=/data/frappe/test \
  --num_epochs=1 \
  --master_resource_request="cpu=0.2,memory=1024Mi" \
  --master_resource_limit="cpu=1,memory=2048Mi" \
  --worker_resource_request="cpu=0.4,memory=1024Mi" \
  --worker_resource_limit="cpu=1,memory=2048Mi" \
  --ps_resource_request="cpu=0.2,memory=1024Mi" \
  --ps_resource_limit="cpu=1,memory=2048Mi" \
  --minibatch_size=64 \
  --num_minibatches_per_task=2 \
  --num_ps_pods=1 \
  --num_workers=1 \
  --evaluation_steps=1 \
  --grads_to_wait=1 \
  --use_async=True \
  --job_name=test-train \
  --log_level=INFO \
  --distribution_strategy=ParameterServerStrategy \
  --image_pull_policy=Never \
  --output=model_output
```

- before wrapper:

```
[2019-12-16 08:38:17,727] [INFO] [worker.py:793:_process_minibatch] Loss is 0.616818368434906
[2019-12-16 08:38:17,743] [INFO] [worker.py:817:_process_eval_task] the evaluation task_id: 398
[2019-12-16 08:38:18,092] [INFO] [worker.py:1012:_evaluate_only] Memory : 521105408
[2019-12-16 08:38:18,114] [INFO] [worker.py:817:_process_eval_task] the evaluation task_id: 400
[2019-12-16 08:38:18,661] [INFO] [worker.py:1012:_evaluate_only] Memory : 521375744
[2019-12-16 08:38:18,684] [INFO] [worker.py:817:_process_eval_task] the evaluation task_id: 401
[2019-12-16 08:38:18,967] [INFO] [worker.py:1012:_evaluate_only] Memory : 521490432
[2019-12-16 08:38:18,968] [INFO] [worker.py:817:_process_eval_task] the evaluation task_id: 402
[2019-12-16 08:38:19,402] [INFO] [worker.py:1012:_evaluate_only] Memory : 521719808
[2019-12-16 08:38:19,412] [INFO] [worker.py:817:_process_eval_task] the evaluation task_id: 403
[2019-12-16 08:38:19,707] [INFO] [worker.py:1012:_evaluate_only] Memory : 522465280
[2019-12-16 08:38:19,708] [INFO] [worker.py:817:_process_eval_task] the evaluation task_id: 404
[2019-12-16 08:38:20,021] [INFO] [worker.py:1012:_evaluate_only] Memory : 522735616
[2019-12-16 08:38:20,023] [INFO] [worker.py:817:_process_eval_task] the evaluation task_id: 405
[2019-12-16 08:38:20,343] [INFO] [worker.py:1012:_evaluate_only] Memory : 523436032
[2019-12-16 08:38:20,346] [INFO] [worker.py:817:_process_eval_task] the evaluation task_id: 406
[2019-12-16 08:38:20,515] [INFO] [worker.py:1012:_evaluate_only] Memory : 523702272
[2019-12-16 08:38:20,517] [INFO] [worker.py:817:_process_eval_task] the evaluation task_id: 407
[2019-12-16 08:38:20,920] [INFO] [worker.py:1012:_evaluate_only] Memory : 524242944
[2019-12-16 08:38:20,922] [INFO] [worker.py:817:_process_eval_task] the evaluation task_id: 408
[2019-12-16 08:38:21,212] [INFO] [worker.py:1012:_evaluate_only] Memory : 524787712
```

- After wrapper

```
[2019-12-16 08:31:54,292] [INFO] [worker.py:793:_process_minibatch] Loss is 0.6835707426071167
[2019-12-16 08:31:54,332] [INFO] [worker.py:817:_process_eval_task] the evaluation task_id: 398
[2019-12-16 08:31:54,971] [INFO] [worker.py:1012:_evaluate_only] Memory : 346025984
[2019-12-16 08:31:54,973] [INFO] [worker.py:817:_process_eval_task] the evaluation task_id: 399
[2019-12-16 08:31:56,401] [INFO] [worker.py:1012:_evaluate_only] Memory : 346025984
[2019-12-16 08:31:56,408] [INFO] [worker.py:817:_process_eval_task] the evaluation task_id: 400
[2019-12-16 08:31:57,805] [INFO] [worker.py:1012:_evaluate_only] Memory : 346025984
[2019-12-16 08:31:57,809] [INFO] [worker.py:817:_process_eval_task] the evaluation task_id: 401
[2019-12-16 08:31:58,379] [INFO] [worker.py:1012:_evaluate_only] Memory : 346025984
[2019-12-16 08:31:58,380] [INFO] [worker.py:817:_process_eval_task] the evaluation task_id: 402
[2019-12-16 08:31:58,962] [INFO] [worker.py:1012:_evaluate_only] Memory : 346025984
[2019-12-16 08:31:58,964] [INFO] [worker.py:817:_process_eval_task] the evaluation task_id: 403
[2019-12-16 08:31:59,515] [INFO] [worker.py:1012:_evaluate_only] Memory : 346025984
[2019-12-16 08:31:59,518] [INFO] [worker.py:817:_process_eval_task] the evaluation task_id: 404
[2019-12-16 08:32:00,085] [INFO] [worker.py:1012:_evaluate_only] Memory : 346025984
[2019-12-16 08:32:00,087] [INFO] [worker.py:817:_process_eval_task] the evaluation task_id: 405
[2019-12-16 08:32:00,554] [INFO] [worker.py:1012:_evaluate_only] Memory : 346025984
[2019-12-16 08:32:00,584] [INFO] [worker.py:817:_process_eval_task] the evaluation task_id: 406
[2019-12-16 08:32:01,389] [INFO] [worker.py:1012:_evaluate_only] Memory : 346025984
[2019-12-16 08:32:01,393] [INFO] [worker.py:817:_process_eval_task] the evaluation task_id: 407
[2019-12-16 08:32:01,807] [INFO] [worker.py:1012:_evaluate_only] Memory : 346025984
[2019-12-16 08:32:01,809] [INFO] [worker.py:817:_process_eval_task] the evaluation task_id: 408
```

### CODE REVIEW

**Comment by skydoorkai** [comment]
Can we also use `tf.function` for the training dataset? If so, we can reuse `create_dataset` method.

File: elasticdl/python/worker/worker.py [hunk 1]
```diff
@@ -814,16 +814,28 @@ def _process_eval_task(self, task):
...
+            eval_dataset = self._dataset_fn(
+                eval_dataset,
+                Mode.EVALUATION,
+                self._task_data_service.data_reader.metadata,
+            )
+            eval_dataset = eval_dataset.batch(self._minibatch_size).prefetch(1)
+            return eval_dataset
+
+        with tf.device("/device:cpu:0"):
+            eval_dataset = create_dataset()
```
Comment (brightcoder01) [inline elasticdl/python/worker/worker.py:836]: From the logic, the gen is either none or a single instance. I think we can just create a single eval_dataset instance from gen and assign it as worker's member variable => self.eval_dataset.
In this way, we have just one eval_dataset instance.

Comment (workingloong) [reply elasticdl/python/worker/worker.py:836]: If we only use a single dataset, we need to determine when the evaluation tasks finish in the dataset pipeline and then continue to execute training tasks.

File: elasticdl/python/worker/task_data_service.py [hunk 2]
```diff
@@ -100,19 +100,22 @@ def report_record_done(self, count, err_msg=""):
...
                     self._current_task = self._pending_tasks[0]
 
-    def get_validation_dataset(self, eval_task):
+    def get_validation_dataset_gen(self, eval_task):
         """
-        If an evaluation task exists, this creates a `tf.data.Dataset`
-        object as well as its corresponding model version and task_id.
-        Otherwise, this returns `None`.
+        If an evaluation task exists, this creates a generator for further
+        creating `tf.data.Dataset`
```
Comment (terrytangyuan) [inline elasticdl/python/worker/task_data_service.py]: What does "further creating `tf.data.Dataset`" mean?

File: elasticdl/python/worker/task_data_service.py [hunk 3]
```diff
@@ -100,19 +100,22 @@ def report_record_done(self, count, err_msg=""):
...
         if not eval_task:
             return None
-        return (
-            create_dataset_from_tasks([eval_task], self.data_reader),
-            eval_task.model_version,
-            eval_task.task_id,
-        )
+        tasks = [eval_task]
+
+        def gen():
```
Comment (terrytangyuan) [inline elasticdl/python/worker/task_data_service.py:111]: So `create_dataset_from_tasks` is not used anymore? Should we remove it then?

File: elasticdl/python/worker/worker.py (hunk 1 above)
Comment (brightcoder01) [reply elasticdl/python/worker/worker.py:836]: I see, thanks

File: elasticdl/python/worker/task_data_service.py (hunk 2 above)
Comment (QiJune) [reply elasticdl/python/worker/task_data_service.py]: The generator could be used to create a dataset instance. I will polish the comment.

File: elasticdl/python/worker/task_data_service.py (hunk 3 above)
Comment (QiJune) [reply elasticdl/python/worker/task_data_service.py:111]: Yes, remove it.

**Comment by QiJune** [comment]
@skydoorkai 
I find that if I move `create_dataset` out, it has to accept an argument, `gen`.

```python
@tf.function(experimental_relax_shapes=True)
    def create_dataset(self, gen):
        eval_dataset = tf.data.Dataset.from_generator(
            gen, self._task_data_service.data_reader.records_output_types
        )
        eval_dataset = self._dataset_fn(
            eval_dataset,
            Mode.EVALUATION,
            self._task_data_service.data_reader.metadata,
        )
        eval_dataset = eval_dataset.batch(self._minibatch_size).prefetch(1)
        return eval_dataset
```

It gets such warnings:

```
[2019-12-17 02:08:41,644] [INFO] [worker.py:829:_process_eval_task] the evaluation task_id: 97
WARNING:tensorflow:11 out of the last 11 calls to <function Worker.create_dataset at 0x7f0eb4672a60> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings is likely due to passing python objects instead of tensors. Also, tf.function has experimental_relax_shapes=True option that relaxes argument shapes that can avoid unnecessary retracing. Please refer to https://www.tensorflow.org/beta/tutorials/eager/tf_function#python_or_tensor_args and https://www.tensorflow.org/api_docs/python/tf/function for more details.
```

Since we only create the training dataset once, we could keep the training dataset logic as former.

**Review by skydoorkai — APPROVED** [review]

**Pull request merged**